### PR TITLE
[PyTorch] Replace cached thread_locals with stack allocation in AOTI

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113578
* #113577
* #112800
* #112799
* #112798
* #112405
* #112174
* __->__ #112116

This changes cached thread_local tensors to stack-allocated buffers. Since we were incidentally caching output in a thread_local, I had to add manual thread_local caching of outputs, which I implemented by caching a buffer and a Tensor whose storage is that buffer and then just memcpying the result into the cached buffer every time. Ideally, memory planning would be able to identify allocations that are the backing storage for outputs, but this should be good enough in the absence of planning.

Differential Revision: [D50416438](https://our.internmc.facebook.com/intern/diff/D50416438/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
Test failures?

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -325,8 +338,9 @@ def __init__(self):
         self.last_seen_device_guard_index = None
         self.supports_intermediate_hooks = True
         self.expr_printer = pexpr
-        self.cached_thread_locals = set()
         self.user_defined_kernel_count = 0
+        self.allow_stack_allocation = None
```
Comment (chenyang78) [inline torch/_inductor/codegen/wrapper.py:342]: Do we initialize ```allow_stack_allocation``` with ```None``` on purpose? If so, wondering what the difference would be between ```None``` and ```False``` for this flag.

**Comment by chenyang78** [comment]
LGTM overall. Left a minor comment. There are some test failures that might be related to this change?

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (swolchok) [reply torch/_inductor/codegen/wrapper.py:342]: None is "not decided yet", False is "decided false"

**Comment by swolchok** [comment]
> Test failures?

the broken test is disabled in fbcode. I'll try a blind fix but if that doesn't work I have no way to debug it.

**Comment by swolchok** [comment]
lintrunner failure is clearly not related to this diff

**Comment by jansel** [comment]
> lintrunner failure is clearly not related to this diff

Lint failure looks related to me.
```
>>> Lint for torch/_inductor/ir.py:
2023-10-31T00:00:01.7590302Z 
2023-10-31T00:00:01.7590494Z   Warning (UFMT) format
2023-10-31T00:00:01.7591001Z     Run `lintrunner -a` to apply this patch.
2023-10-31T00:00:01.7591306Z 
2023-10-31T00:00:01.7591590Z     You can run `lintrunner -a` to apply this patch.
2023-10-31T00:00:01.7591963Z 
2023-10-31T00:00:01.7592288Z     5232  5232 |         algorithm,
2023-10-31T00:00:01.7592834Z     5233  5233 |     ):
2023-10-31T00:00:01.7593433Z     5234  5234 |         transposed = True
2023-10-31T00:00:01.7594388Z     5234       |-        (inputs, constant_args, kernel_layout, _,) = _prepare_convolution_fusion_create(
2023-10-31T00:00:01.7595248Z           5235 |+        (
2023-10-31T00:00:01.7595820Z           5236 |+            inputs,
2023-10-31T00:00:01.7596438Z           5237 |+            constant_args,
2023-10-31T00:00:01.7597088Z           5238 |+            kernel_layout,
2023-10-31T00:00:01.7597685Z           5239 |+            _,
2023-10-31T00:00:01.7598389Z           5240 |+        ) = _prepare_convolution_fusion_create(
2023-10-31T00:00:01.7599088Z     5236  5241 |             cls,
2023-10-31T00:00:01.7599658Z     5237  5242 |             x,
2023-10-31T00:00:01.7600245Z     5238  5243 |             weight,
```

**Review by desertfire — COMMENTED** [review]
The coding LGTM. Left some minor comment. A more general concern I have is this adds more complexity to the wrapper codegen. With the memory planning PR https://github.com/pytorch/pytorch/pull/112178 merged (reverted with some ROCM test failures to be fixed/skiped), should we give that a try to see it is sufficient enough for your use case?

**Comment by desertfire** [comment]
Also I left some comments in [D50416438](https://our.internmc.facebook.com/intern/diff/D50416438/) for fixing lint issues.

**Comment by swolchok** [comment]
> The coding LGTM. Left some minor comment. A more general concern I have is this adds more complexity to the wrapper codegen. With the memory planning PR #112178 merged (reverted with some ROCM test failures to be fixed/skiped), should we give that a try to see it is sufficient enough for your use case?

My use case is sensitive to the overhead of individual aoti_torch calls, to say nothing of heap allocations. We will notice the cost of `aoti_torch__alloc_from_pool` and in general the at::Tensor API that the memory planning PR continues to use. (I'd run it to double-check, but there's no fbcode diff.) We could investigate migrating the memory planner to use ArrayRefTensor (see later diffs in this stack) and a simple bump-the-pointer allocator, at which point it might work, but the existence of a potential better future solution doesn't seem like a good reason to block a cheap good solution from landing.

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -2170,6 +2170,15 @@ def constant_to_device(self, device):
...
 
 
+def is_contiguous_strides_for_shape(stride, shape):
+    return all(
+        size == 1 or left == right
+        for left, right, size in zip(
+            stride, FlexibleLayout.contiguous_strides(shape), shape
+        )
+    )
+
```
Comment (ezyang) [inline torch/_inductor/ir.py:2180]: just refactor right

File: test/inductor/test_aot_inductor.py [hunk 3]
```diff
@@ -1063,6 +1098,7 @@ class AOTInductorTestNonABICompatibleCuda(TestCase):
     abi_compatible = False
     check_model = check_model
     check_model_with_multiple_inputs = check_model_with_multiple_inputs
+    allow_stack_allocation = False
```
Comment (ezyang) [inline test/inductor/test_aot_inductor.py:1101]: So I guess there are some gaps with the implementation here? What are the gaps?

**Comment by ezyang** [comment]
@desertfire you tagged me on this but I have missed some of the original changes that got us into this state. In particular, why were we caching thread locals in the first place?

**Comment by swolchok** [comment]
> In particular, why were we caching thread locals in the first place?

https://github.com/pytorch/pytorch/pull/110892

File: test/inductor/test_aot_inductor.py (hunk 3 above)
Comment (swolchok) [reply test/inductor/test_aot_inductor.py:1101]: stack-allocating CUDA tensors does not make sense -- they need to be in GPU memory. If we want to add a test in CUDA mode with the flag on, we could do that, but the default test case should have it off IMO.

File: torch/_inductor/ir.py (hunk 2 above)
Comment (swolchok) [reply torch/_inductor/ir.py:2180]: yep, needed to call this

File: test/inductor/test_aot_inductor.py (hunk 3 above)
Comment (ezyang) [reply test/inductor/test_aot_inductor.py:1101]: !!! you're stack allocating the tensor data, not the tensor impl, ok, makes more sense lol

**Review by desertfire — APPROVED** [review]

**Comment by desertfire** [comment]
> @desertfire you tagged me on this but I have missed some of the original changes that got us into this state. In particular, why were we caching thread locals in the first place?

As @swolchok pointed out, thread locals were introduced to reduce tensor allocation overhead for cpu models. This PR is fine given it is improving what was already there and does not really affect GPU models.

**Review by jansel — APPROVED** [review]
Fix lint failure

**Comment by swolchok** [comment]
> memory planning PR

I double-checked this; enabling memory planning instead of stack allocation causes a 3x slowdown in our internal benchmark.

**Pull request closed without merging**